src/pmdas/bpf/bpf.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * BPF wrapper metric module.
+ *
+ * Copyright (c) 2021 Netflix, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+#include <pcp/pmapi.h>
+#include <pcp/pmda.h>
+#include "domain.h"
+#include "modules/module.h"
+#include <sys/stat.h>
+#include <sys/resource.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include <dlfcn.h>
+#include "bpf.h"
+
+static int	isDSO = 1;		/* =0 I am a daemon */
+static char	mypath[MAXPATHLEN];
+
+/* metric and indom configuration will be dynamically filled in by modules */
+static pmdaMetric * metrictab;
+static pmdaIndom * indomtab;
+static int metric_count = 0;
+static int indom_count = 0;
+
+/* all modules collected here (whether initialised or not) */
+static module** modules;
+static int module_count;
+
+/* command line option handling - both short and long options */
+static pmLongOptions longopts[] = {
+    PMDA_OPTIONS_HEADER("Options"),
+    PMOPT_DEBUG,
+    PMDAOPT_DOMAIN,
+    PMDAOPT_LOGFILE,
+    PMOPT_HELP,
+    PMDA_OPTIONS_TEXT("\nExactly one of the following options may appear:"),
+    PMDAOPT_INET,
+    PMDAOPT_PIPE,
+    PMDAOPT_UNIX,
+    PMDAOPT_IPV6,
+    PMDA_OPTIONS_END
+};
+static pmdaOptions opts = {
+    .short_options = "D:d:i:l:pu:6:?",
+    .long_options = longopts,
+};
+
+/*
+ * callback provided to pmdaFetch
+ */
+static int
+bpf_fetchCallBack(pmdaMetric *mdesc, unsigned int inst, pmAtomValue *atom)
+{
+    unsigned int	cluster = pmID_cluster(mdesc->m_desc.pmid);
+    unsigned int	item = pmID_item(mdesc->m_desc.pmid);
+
+    // find module for this cluster, and issue fetch to module
+    // if somehow this bypasses PMCD PMNS filtering it is possible that
+    // a call to an uninitialised module occurs; the module is to handle this
+    for (int i = 0; i < module_count; i++)
```

Comment:
could be a dict/hash, but the real-world performance gain is probably negligible as there won't be 1000s of modules

Reply:
I initially had this mapped as a direct cluster id -> module mapping, which made it a linear lookup, but changed this when introducing dynamic loading. What I will do is add a post-configuration cluster id -> module setup. Thanks!

Reply:
fixed in 8d708ceb7